Fix/improve how hostname, ip and port are handled ( CFE-953 CFE-721 CFE-385 CFE-2481 )

olehermanse opened the pull request:
This is a bigger pull request which addresses multiple issues with the hostname and custom port features (CFE-953 CFE-546).

Previous PR for reference: cfengine/core#2699

Must be merged together with Pull Requests in masterfiles and enterprise.

olehermanse commented:
Last thing missing is POLICY_SERVER global should be bootstrap arg (host:port)

olehermanse commented:
POLICY_SERVER is now longer and contains the same as bootstrap arg and policy_server.dat
(host:port)

olehermanse commented:
This PR makes both the c code and variables a lot more clear imo.
Also, bootstrapping now shows the complete host port string:
```
 notice: Bootstrap to 'olehub:5309' completed successfully!
```

olehermanse commented:
Do not merge yet, I am still testing.

olehermanse commented:
Warning: This definitely does not work as of now, because POLICY_SERVER is used internally for copy_from files promises. I am working on a fix.

olehermanse commented on cf-agent/cf-agent.c:
```diff
@@ -1981,6 +1981,8 @@ static int AutomaticBootstrap(GenericAgentConfig *config)
         {
             config->agent_specific.agent.bootstrap_argument = xstrdup(ipaddr);
+            config->agent_specific.agent.bootstrap_ip       = xstrdup(ipaddr);
+            config->agent_specific.agent.bootstrap_host     = xstrdup(ipaddr);
```
I don't know how to test the avahi stuff, but I populated the variables to be more similar to a regular bootstrap.

olehermanse replied:
I don't know how to test the avahi stuff, but I populated the variables to be more similar to a regular bootstrap.

olehermanse replied:
I don't know how to test the avahi stuff, but I populated the variables to be more similar to a regular bootstrap.

olehermanse commented:
Do not merge yet. I should add the new sys.policy_hub variable before we merge, to ensure old stuff doesn't break.

olehermanse commented:
https://github.com/cfengine/core/pull/2727/commits/58b073e06605dc0a2dce028455e3d6dc05f51ca4 <- Did I do this correctly with regards to Changelog?

olehermanse commented:
Here are my instructions for usage/testing:
https://gist.github.com/olehermanse/d98b5b1e0480701db73efb75aa150dee

olehermanse commented:
I've tested this and it seems like it's working fine. The only issue I've found is: https://tracker.mender.io/browse/CFE-2494 But I don't think that has anything to do with this patch.

olehermanse commented:
@kacf @jimis Please review. This is a bigger PR, but should be easy to understand commit by commit.

Must be merged with: 
https://github.com/cfengine/masterfiles/pull/824
https://github.com/cfengine/enterprise/pull/324

kacf commented:
Since this is network code I will let @jimis review this.

jimis commented:
Interesting, but I believe it will definitely cause breakage and grief. What is the rationale behind changing the "policy_hub" variable? Can we follow a more backwards-compatible way? I think it would be nice to reopen the ticket and keep track of this.

olehermanse commented:
(Discussed with jimis:)

sys.policy_hub and POLICY_SERVER are both exactly like before when not using the new host:port feature. Backwards compatibility should only be an issue with old policy when using the new bootstrap features.

olehermanse commented:
I think this PR makes the variables a lot more consistent, with policy_server.dat, POLICY_SERVER and sys.policy_hub now being exactly the same. It also improves log output, for example here:
```
  notice: Bootstrap to 'olehub:5308' completed successfully!
```

olehermanse commented:
@jimis: Discussed with @nickanderson and it seems to make sense to instead do this on policy side:
`policy_hub` can be either host or ip
`policy_hub_ip` is always ip
`policy_hub_port` is always port

My idea was to have policy_hub be exactly what you bootstrapped to. Having policy_hub variable include port number is not useful according to nick.

Then I should be able to fix the ACL issue for `host/16` : https://tracker.mender.io/browse/CFE-2495

jimis commented:
> @jimis: Discussed with @nickanderson and it seems to make sense to instead do this on policy side:
> policy_hub can be either host or ip
> policy_hub_ip is always ip
> policy_hub_port is always port

Fully agree, as we discussed extensively.

Regarding the `POLICY_SERVER` global, it's better to work either as host or ip, and keep it what we bootstrapped to. Whether it has the port or not is not clear yet, you'll figure it out. :+1:

olehermanse commented:
I've now changed the policy_hub variable as discussed, rebased and pushed again. 
`sys.policy_hub` is now host or ip (not port).

jimis commented:
Then the next step is to build in Jenkins' pull request analyzer!

jimis commented on contrib/masterfiles/git-failsafe.cf:
```diff
@@ -198,7 +198,7 @@ bundle agent git_update
         comment => "Check whether a validation stamp is available for a new policy update to reduce the distributed load",
          handle => "check_valid_update",
-      copy_from => u_rcp("$(master_location)/cf_promises_validated","$(sys.policy_hub)"),
+      copy_from => u_rcp("$(master_location)/cf_promises_validated","$(sys.policy_hub_ip)"),
```
Most `$(sys.policy_hub)` in the policy would be best to remain unchanged, since the code handles both IP and hostnames (especially true after the ACLs are fixed).

Just mentioning here to make sure we're on the same ship. :-)

olehermanse replied:
Yes I agree, however I think maybe doing it like this for now, and changing back once ACL's are fixed is better(?).

jimis replied:
Sounds good! Then just leave the entries that are relevant to ACLs, and change back the `copy_from` ones to `policy_hub`.

olehermanse commented:
Must be merged with: 
https://github.com/cfengine/enterprise/pull/324
https://github.com/cfengine/nova/pull/990

olehermanse commented:
I've rebase and squashed the commits into a few that make sense on their own. I've started a jenkins build and will test this and hope we can merge if it looks promising.

Remaining:
1. Build & Testing
2. Fix ACL issues with hostname
3. Move away from using sys.policy_hub_ip where not strictly needed.

@jimis Please review.

Must be merged with: 
cfengine/enterprise#324
cfengine/nova#990

Relevant jenkins build:
https://ci.cfengine.com/job/bootstrap-enterprise-pr/82/

olehermanse commented:
Ideas for improving policyserver.c performance:
1. When policy server is ip - Leave host as NULL and avoid alloc and lookup. 
2. Cache IP with a time limit to avoid many lookups. Cache can last **long** as a host changing ip should not happen too frequently. 

I really like that we can do these sort of changes while keeping the interface (get methods) the same.

jimis commented on libpromises/policyserver.c:
```diff
@@ -0,0 +1,300 @@
+    char *filename = PolicyServerFilename(workdir);
+
+    FILE *file = fopen(filename, "w");
+    if (!file)
```
Avoid `!file` since file is no `bool`, prefer `file != NULL`

jimis commented on libpromises/policyserver.c:
```diff
@@ -0,0 +1,300 @@
+        return false;
+    }
+
+    fprintf(file, "%s", new_policy_server);
```
Since you are at it, would you mind adding a trailing `\n` in a separate commit at the end? I don't think it will break existing code, and I've been wanting to do it a since a long time...

jimis commented on libpromises/policyserver.c:
```diff
@@ -0,0 +1,300 @@
+//*******************************************************************
+
+// TODO: Shouldn't be necessary? Can be inline too
+static const char* PolicyServerCheckedReturn(char* rval)
```
Don't use the word `rval` for a string, `Rval` is a very widely type used in the CFEngine source code so it creates confusion. Use `host` or `ip` or whatever this is, or even `s` is better in this context since it's simple.

jimis commented on libpromises/policyserver.c:
```diff
@@ -0,0 +1,300 @@
+// POLICY SERVER GET FUNCTIONS:
+//*******************************************************************
+
+// TODO: Shouldn't be necessary? Can be inline too
```
No it's not necessary, in fact it's quite confusing since it's unclear what is checked via the parameter and what via the global.

jimis commented on libpromises/policyserver.h:
```diff
@@ -0,0 +1,131 @@
+*/
+
+/** @file
+ * @brief Replaces global strings for POLICY_SERVER with get and set methods
```
It's nice that you write a global purpose description in the header file. However it should mean something to someone reading it, without knowing what's in the file or the past that this file replaces. 

For example: Functions to set the policy server by using IP/hostname/port, and to query it back.

jimis commented on libpromises/policyserver.c:
```diff
@@ -0,0 +1,300 @@
+
+void SetPolicyServerFromFile(EvalContext *ctx, const char* workdir)
+{
+    char* contents = ReadPolicyServerFile(workdir);
```
Star to the right, here and in other places.

jimis commented on libpromises/policyserver.c:
```diff
@@ -0,0 +1,300 @@
+// POLICY SERVER VARIABLES:
+//*******************************************************************
+
+static char *POLICY_SERVER      = NULL;
```
Add a comment that `POLICY_SERVER` is the exact string to the one we bootstrapped to and can be IP or hostname and includes port. And another comment that `POLICY_SERVER_HOST` has the port removed, and can be host or IP. If I undestand this right.

jimis commented on libpromises/policyserver.c:
```diff
@@ -0,0 +1,300 @@
+        // Parse policy server in a separate buffer:
+        char *host, *port;
+        char buffer[CF_MAX_SERVER_LEN];
+        strncpy(buffer, new_policy_server, CF_MAX_SERVER_LEN);
```
`strncpy()` is seldom needed, it's not a good function. Just allocate here, since setting this is not frequent operation.

jimis commented:
> Ideas for improving policyserver.c performance:
> 1. When policy server is ip - Leave host as NULL and avoid alloc and lookup. 

This indeed sounds necessary!

olehermanse commented:
@jimis I think I have addressed all the issues you commented and we discussed.

The plan going forward is:
1. Build and test in travis + jenkins.
2. Get this merged
3. Write unit tests
4. Fix ACL Issue

olehermanse commented:
Done with unit tests.  Waiting for travis to see if it's still failing the one acceptance test. Onto jenkins build if travis succeeds.

olehermanse commented:
Started build in jenkins: 
https://ci.cfengine.com/job/bootstrap-enterprise-pr/87/
https://ci.cfengine.com/job/testing-enterprise-pr/75/

jimis commented on cf-agent/cf-agent.c:
```diff
@@ -1852,7 +1853,8 @@ static PromiseResult ParallelFindAndVerifyFilesPromises(EvalContext *ctx, const
 static bool VerifyBootstrap(void)
 {
-    if (NULL_OR_EMPTY(POLICY_SERVER))
+    const char* policy_server = PolicyServerGet();
```
style

jimis commented on libpromises/addr_lib.c:
```diff
@@ -566,6 +566,53 @@ bool IsInterfaceAddress(const Item *ip_addresses, const char *adr)
+    }
+
+    // Leading whitespace:
+    while ( isspace(s[0]) )
```
In general, we don't use whitespace around the expression unless needed.

It's OK when it enhances readability, for example with complex expressions or for alignment purposes.

jimis commented on libpromises/bootstrap.c:
```diff
@@ -75,159 +73,165 @@ bool BootstrapAllowed(void)
     }
+    if (fgets(contents, CF_MAX_SERVER_LEN, fp) == NULL)
+    {
+        fclose(fp);
```
A verbose message here would help debugging in cases this happens.

jimis commented on libpromises/evalfunction.c:
```diff
@@ -24,6 +24,7 @@
 
 #include <evalfunction.h>
 
+#include <policy_server.h>
```
you are including it twice

jimis commented on libpromises/policy_server.c:
```diff
@@ -0,0 +1,182 @@
+
+/**
+ * @brief   Gets the port part of the policy server.
+ * @return  Read-only null terminated string of port number.
```
Can  you mention in the doc comment that NULL is an acceptable value, if the port is not explicitly set? Same with the rest of the functions that this stands.

jimis commented on libpromises/policy_server.h:
```diff
@@ -0,0 +1,72 @@
+ * @param ctx EvalContext is used to set related variables
+ * @param new_policy_server can be 'host:port', same as policy_server.dat
+ */
+void PolicyServerSet(const char *new_policy_server);
```
Don't duplicate the doc comments, having them in the implementation file is enough.

jimis commented:
Well, jenkins is red unfortunately: https://ci.cfengine.com/job/testing-enterprise-pr/75/

olehermanse commented:
You linked to the wrong build. As we discussed, it seems builds and unit tests are passing but acceptance tests are failing. Will investigate.

olehermanse commented:
Started new build in Jenkins: 
https://ci.cfengine.com/job/bootstrap-enterprise-pr/92/
https://ci.cfengine.com/job/testing-enterprise-pr/80/

olehermanse commented:
@jimis Green across the board! Please merge. ( See last comment )

jimis commented:
wow this merge brought in several newline changes with CRLF; please disable all CRLF endings in your editor, @olehermanse ! This might be related to the XML illegal characters we've been seeing.

jimis commented:
After much searching, I am not sure I found any CRLFs introduced in non-policy files. And I /think/ (as it's not easy to grep that) all the special characters you introduced are in already DOS-mode files, so nothing wrong with that.

I still get the XML errors though, now I got it on AIX...

olehermanse commented:
Atom uses whatever line ending is already used in the file.

About the errors, I'm not sure what's causing it.
